### PR TITLE
Add NodeletLazy abstract class for lazy transport

### PR DESCRIPTION
**Pull request by wkentaro**
This class is to crete nodelet for 'lazy' transport.
Class for Python node is already merged in kinetic-devel branch of topic_tools.  https://github.com/ros/ros_comm/pull/713

Lazy transport is an essential feature for perception nodelets that handle image and point cloud, because
the processing should run only when it is necessary to reduce the CPU load.

FYI, the similar nodelet abstract class is used in ros-perception/opencv_apps package. https://github.com/ros-perception/opencv_apps/blob/indigo/include/opencv_apps/nodelet.h

### CODE REVIEW

**Comment by wkentaro** [comment]
Maintainers, could you please review this?
Possibly, the first step to view how it works is below:

``` bash
cd test_nodelet_topic_tools
catkin build --this
rostest test/test_nodelet_lazy.launch --text
```

**Comment by wkentaro** [comment]
Any comments, please/

**Comment by mikaelarguedas** [comment]
sorry for the late reply, will review this shortly.
Thanks!

**Comment by wkentaro** [comment]
Friendly ping.

**Comment by wkentaro** [comment]
ping

**Comment by mikaelarguedas** [comment]
finally got back decent internet, will review this today, sorry for the very long wait

File: nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    // option for logging about being subscribed
+    pnh_->param("verbose_connection", verbose_connection_, false);
+    if (!verbose_connection_)
+    {
+      nh_->param("verbose_connection", verbose_connection_, false);
+    }
+    // timer to warn when no connection in a few seconds
+    ever_subscribed_ = false;
+    timer_ever_subscribed_ = nh_->createWallTimer(
+      ros::WallDuration(5),
```
Comment (mikaelarguedas) [inline nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h]: Would it make sense to expose this timer as a ros param ?

File: nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    * Post processing of initialization of nodelet.
+    * You need to call this method in order to use always_subscribe
+    * feature.
+    */
+  virtual void onInitPostProcess()
+  {
+    if (!lazy_)
+    {
+      boost::mutex::scoped_lock lock(connection_mutex_);
+      ever_subscribed_ = true;
```
Comment (mikaelarguedas) [inline nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h]: set flag after subscribe succeeded

File: nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    {
+      boost::mutex::scoped_lock lock(connection_mutex_);
+      for (size_t i = 0; i < publishers_.size(); i++)
+      {
+        ros::Publisher pub = publishers_[i];
+        if (pub.getNumSubscribers() > 0)
+        {
+          if (!ever_subscribed_)
+          {
+            ever_subscribed_ = true;
```
Comment (mikaelarguedas) [inline nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h]: same as above

File: nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  /** @brief
+    * This method is called when publisher is unsubscribed by other
+    * nodes.
+    * Shut down subscribers in this method.
+    */
+  virtual void unsubscribe() = 0;
+
+  /** @brief
+    * Advertise a topic and watch the publisher. Publishers which are
+    * created by this method.
```
Comment (mikaelarguedas) [inline nodelet_topic_tools/include/nodelet_topic_tools/nodelet_lazy.h]: Rephrase this as a single sentence? the meaning of "Publishers which are created by this method" is unclear to me. Did you mean "update the list of Publishers created by this method" ?

**Review by mikaelarguedas — CHANGES_REQUESTED** [review]

**Comment by mikaelarguedas** [comment]
@wkentaro Code looks good to me overall, a few nicktip comments to address/clarify but all of them are cosmetics.

**Comment by wkentaro** [comment]
Thank you for the reviews. I fixed the code.

**Comment by mikaelarguedas** [comment]
great! thanks for iterating on this

**Review by mikaelarguedas — APPROVED** [review]

**Comment by wkentaro** [comment]
Thanks!
